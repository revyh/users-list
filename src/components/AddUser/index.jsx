import React from 'react';
import FloatingActionButton from 'material-ui/FloatingActionButton';
import Dialog from 'material-ui/Dialog';
import FlatButton from 'material-ui/FlatButton';
import TextField from 'material-ui/TextField';
import AddIcon from 'material-ui/svg-icons/content/add';
import RequiredMark from 'components/RequiredMark';

export default class AddUser extends React.Component {
  state = {
    isOpen: false,
    firstName: '',
    lastName: '',
    department: '',
  };

  handleOpen = () => {
    this.setState({isOpen: true});
  };

  handleClose = () => {
    this.setState({isOpen: false});
  };

  handleAdd = () => {
    const {onAdd} = this.props;
    const {firstName, lastName, department} = this.state;

    onAdd && onAdd(firstName, lastName, department);
    this.setState({
      isOpen: false,
      firstName: '',
      lastName: '',
      department: '',
    });
  };

  handleFirstNameChange = ev => {
    this.setState({firstName: ev.target.value});
  };

  handleLastNameChange = ev => {
    this.setState({lastName: ev.target.value});
  };

  handleDepartmentChange = ev => {
    this.setState({department: ev.target.value});
  };

  render() {
    const {style, getErrors} = this.props;
    const {firstName, lastName, department, isOpen} = this.state;

    const actions = [
      <FlatButton
        key="cancel"
        label="Cancel"
        primary
        onTouchTap={this.handleClose}
      />,
      <FlatButton
        key="add"
        label="Add"
        primary
        disabled={!!getErrors && !!getErrors(firstName, lastName)}
        onTouchTap={this.handleAdd}
      />,
    ];

    return (
      <div style={style}>
        <FloatingActionButton onTouchTap={this.handleOpen}>
          <AddIcon />
        </FloatingActionButton>
        <Dialog
          title="Add user"
          actions={actions}
          modal={false}
          open={isOpen}
          onRequestClose={this.handleClose}
        >
          <TextField
            key="firstName"
            floatingLabelText={
              <span>First name <RequiredMark /></span>
            }
            value={firstName}
            fullWidth
            keyboardFocused
            onChange={this.handleFirstNameChange}
          />
          <TextField
            key="lastName"
            floatingLabelText={
              <span>Last name <RequiredMark /></span>
            }
            value={lastName}
            fullWidth
            onChange={this.handleLastNameChange}
          />
          <TextField
            key="department"
            floatingLabelText="department"
            value={department}
            fullWidth
            onChange={this.handleDepartmentChange}
          />
        </Dialog>
      </div>
    );
  }
}
