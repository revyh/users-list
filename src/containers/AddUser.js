import {connect} from 'react-redux';
import {defaultMemoize} from 'reselect';
import {actions, selectors} from 'features/users';
import Presentational from 'components/AddUser';

const getErrors = defaultMemoize(selectors.getAddUsersErrors);

function mapDispatchToProps(dispatch) {
  return {
    getErrors,
    onAdd: (...args) => dispatch(actions.addUser(...args)),
  };
}

export default connect(null, mapDispatchToProps)(Presentational);
