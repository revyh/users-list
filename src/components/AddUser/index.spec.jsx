import React from 'react';
import {shallow} from 'enzyme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import AddUser from '.';

function shallowWithContext(node) {
  return shallow(node, {context: {muiTheme: getMuiTheme()}});
}

function getTextField(root, key) {
  return root.findWhere(
    node => node.name() === 'TextField' && node.key() === key,
  );
}

it('should render', () => {
  expect(shallow(<AddUser />)).toMatchSnapshot();
});

it('should open dialog when floating action button clicked', () => {
  const addUser = shallow(<AddUser />);

  addUser.children('FloatingActionButton').simulate('touchTap');

  expect(addUser.children('Dialog').prop('open')).toBeTruthy();
});

it('should close dialog when clicked outside', () => {
  const addUser = shallow(<AddUser />);

  addUser.children('FloatingActionButton').simulate('touchTap');
  addUser.children('Dialog').simulate('requestClose');

  expect(addUser.children('Dialog').prop('open')).toBeFalsy();
});

it('should close dialog when clicked on dialog \'cancel\' button', () => {
  const addUser = shallow(<AddUser />);

  addUser.children('FloatingActionButton').simulate('touchTap');
  shallowWithContext(addUser.children('Dialog').prop('actions')[0])
    .simulate('touchTap');

  expect(addUser.children('Dialog').prop('open')).toBeFalsy();
});

it('should close dialog when clicked on dialog \'add\' button', () => {
  const addUser = shallow(<AddUser />);

  addUser.children('FloatingActionButton').simulate('touchTap');
  shallowWithContext(addUser.children('Dialog').prop('actions')[1])
    .simulate('touchTap');

  expect(addUser.children('Dialog').prop('open')).toBeFalsy();
});

it('should pass firstName and lastName to validation function', () => {
  const mockGetErrors = jest.fn();
  const addUser = shallow(<AddUser getErrors={mockGetErrors} />);

  addUser.children('FloatingActionButton').simulate('touchTap');
  getTextField(addUser, 'firstName')
    .simulate('change', {target: {value: 'some first name'}});
  getTextField(addUser, 'lastName')
    .simulate('change', {target: {value: 'some last name'}});

  expect(mockGetErrors)
    .toHaveBeenLastCalledWith('some first name', 'some last name');
});

it('should disable dialog \'add\' button if data is invalid', () => {
  const addUser = shallow(<AddUser getErrors={jest.fn(() => true)} />);

  addUser.children('FloatingActionButton').simulate('touchTap');
  const button = addUser.children('Dialog').prop('actions')[1];

  expect(shallowWithContext(button).prop('disabled')).toBeTruthy();
});

it('should enable dialog \'add\' button if data is valid', () => {
  const addUser = shallow(<AddUser getErrors={jest.fn(() => false)} />);

  addUser.children('FloatingActionButton').simulate('touchTap');
  const button = addUser.children('Dialog').prop('actions')[1];

  expect(shallowWithContext(button).prop('disabled')).toBeFalsy();
});

it('should trigger \'onAdd\' when clicked on \'add\' button', () => {
  const mockOnAdd = jest.fn();
  const addUser = shallow(<AddUser onAdd={mockOnAdd} />);

  addUser.children('FloatingActionButton').simulate('touchTap');

  getTextField(addUser, 'firstName')
    .simulate('change', {target: {value: 'some first name'}});
  getTextField(addUser, 'lastName')
    .simulate('change', {target: {value: 'some last name'}});
  getTextField(addUser, 'department')
    .simulate('change', {target: {value: 'some department'}});

  shallowWithContext(addUser.children('Dialog').prop('actions')[1])
    .simulate('touchTap');

  expect(mockOnAdd).toHaveBeenCalledWith(
    'some first name',
    'some last name',
    'some department',
  );
});
