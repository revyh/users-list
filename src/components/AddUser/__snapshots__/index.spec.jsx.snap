// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div>
  <FloatingActionButton
    disabled={false}
    mini={false}
    onTouchTap={[Function]}
    secondary={false}
    zDepth={2}
  >
    <ContentAdd />
  </FloatingActionButton>
  <Dialog
    actions={
      Array [
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={true}
          secondary={false}
      />,
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Add"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={true}
          secondary={false}
      />,
      ]
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={false}
    modal={false}
    onRequestClose={[Function]}
    open={false}
    repositionOnUpdate={true}
    title="Add user"
  >
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      floatingLabelText={
        <span>
          First name 
          <RequiredMark />
        </span>
      }
      fullWidth={true}
      keyboardFocused={true}
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      floatingLabelText={
        <span>
          Last name 
          <RequiredMark />
        </span>
      }
      fullWidth={true}
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      floatingLabelText="department"
      fullWidth={true}
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
  </Dialog>
</div>
`;
